protocol/kafka_confluent/v2/protocol.go

```diff
@@ -0,0 +1,225 @@
+
+	// receiver
+	incoming chan *kafka.Message
+}
+
+func New(opts ...Option) (*Protocol, error) {
+	p := &Protocol{
+		consumerPollTimeout: 100,
+		incoming:            make(chan *kafka.Message),
+	}
+	if err := p.applyOptions(opts...); err != nil {
+		return nil, err
+	}
+
+	if p.kafkaConfigMap != nil {
+		if p.consumerTopics != nil && p.consumer == nil {
+			consumer, err := kafka.NewConsumer(p.kafkaConfigMap)
+			if err != nil {
+				return nil, err
+			}
+			p.consumer = consumer
+		}
+		if p.producerDefaultTopic != "" && p.producer == nil {
+			producer, err := kafka.NewProducer(p.kafkaConfigMap)
+			if err != nil {
+				return nil, err
+			}
+			p.producer = producer
+		}
+		if p.producer == nil && p.consumer == nil {
+			return nil, fmt.Errorf("at least receiver or sender topic must be set")
+		}
+	}
+
+	if p.kafkaConfigMap == nil && p.producer == nil && p.consumer == nil {
+		return nil, fmt.Errorf("at least one of the following to initialize the protocol must be set: config, producer, or consumer")
+	}
+	if p.producer != nil {
+		p.producerDeliveryChan = make(chan kafka.Event)
+	}
+	return p, nil
+}
+
+func (p *Protocol) applyOptions(opts ...Option) error {
+	for _, fn := range opts {
+		if err := fn(p); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (p *Protocol) Send(ctx context.Context, in binding.Message, transformers ...binding.Transformer) (err error) {
+	if p.producer == nil {
+		return fmt.Errorf("producer client must be set")
+	}
+	defer in.Finish(err)
+
+	kafkaMsg := &kafka.Message{
+		TopicPartition: kafka.TopicPartition{
+			Topic:     &p.producerDefaultTopic,
+			Partition: kafka.PartitionAny,
+		},
+	}
+
+	if topic := cecontext.TopicFrom(ctx); topic != "" {
+		kafkaMsg.TopicPartition.Topic = &topic
+	}
+
+	if messageKey := MessageKeyFrom(ctx); messageKey != "" {
+		kafkaMsg.Key = []byte(messageKey)
+	}
+
+	err = WriteProducerMessage(ctx, in, kafkaMsg, transformers...)
+	if err != nil {
+		return err
+	}
+
+	err = p.producer.Produce(kafkaMsg, p.producerDeliveryChan)
+	if err != nil {
+		return err
+	}
+	e := <-p.producerDeliveryChan
+	m := e.(*kafka.Message)
+	if m.TopicPartition.Error != nil {
+		return m.TopicPartition.Error
+	}
+	return nil
+}
+
+func (p *Protocol) OpenInbound(ctx context.Context) error {
+	if p.consumer == nil {
+		return fmt.Errorf("the consumer client must be set")
+	}
+	if p.consumerTopics == nil {
+		return fmt.Errorf("the consumer topics must be set")
+	}
+
+	p.consumerMux.Lock()
+	defer p.consumerMux.Unlock()
+	logger := cecontext.LoggerFrom(ctx)
+
+	// Query committed offsets for each partition
+	if positions := TopicPartitionOffsetsFrom(ctx); positions != nil {
+		if err := p.consumer.Assign(positions); err != nil {
+			return err
+		}
+	}
+
+	logger.Infof("Subscribing to topics: %v", p.consumerTopics)
+	err := p.consumer.SubscribeTopics(p.consumerTopics, p.consumerRebalanceCb)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if !p.consumer.IsClosed() {
```

Comment:
use mutex here when removing the "global" mutex above

Reply:
See the [above](https://github.com/cloudevents/sdk-go/pull/988#discussion_r1527525977).